tests/kernel/workq/work_queue_api/src/main.c

```diff
@@ -37,6 +50,154 @@ static struct k_sem sync_sema;
+ * @details To test it we use 2 functions-handlers.
+ * - fifo_work_first() added to the work_item_1 and fifo_work_second()
+ * added to the work_item_2.
+ * - We expect that firstly should run work_item_1, and only then
+ * will run work_item_2.
+ * - To test it, we initialize semaphore sema_fifo_one
+ * with count 1(available) and fifo_work_first() takes that semaphore.
+ * fifo_work_second() is waiting for the semaphore sema_fifo_two,
+ * which will be given from function fifo_work_first().
+ * - If work_item_2() will try to be executed before work_item_1(),
+ * will happen a timeout error.
+ * - Because sema_fifo_two will be never obtained by fifo_work_second()
+ * due to K_FOREVER macro in it while waiting for the semaphore.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_process_work_items_fifo(void)
+{
+	k_work_init(&work_item_1, fifo_work_first);
+	k_work_init(&work_item_2, fifo_work_second);
+
+	/**TESTPOINT: submit work items to the queue in fifo manner*/
+	k_work_submit_to_queue(&workq, &work_item_1);
+	k_work_submit_to_queue(&workq, &work_item_2);
+}
+
+/**
+ * @brief Test kernel support scheduling work item that is to be processed
+ * after user defined period of time
+ * @details
+ * - For that test is using semaphore sync_sema, with initial count 0.
+ * - In that test we measure actual spent time and compare it with time
+ * which was measured by function k_delayed_work_remaining_get().
+ * - Using system clocks we measure actual spent time
+ * in the period between delayed work submitted and delayed work
+ * executed.
+ * - To know that delayed work was executed, we use semaphore.
+ * - Right after semaphore was given from handler function, we stop
+ * measuring actual time.
+ * - Then compare results.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_sched_delayed_work_item(void)
+{
+	s32_t ms_remain, ms_spent, start_time, stop_time, cycles_spent;
+
+	s32_t ms_delta = 10;
+
+	k_sem_reset(&sync_sema);
+
+	/* TESTPOINT: init delayed work to be processed */
+	/* only after specific period of time */
+	k_delayed_work_init(&work_item_delayed, common_work_handler);
+	start_time = k_cycle_get_32();
+	k_delayed_work_submit_to_queue(&workq, &work_item_delayed, TIMEOUT);
+	ms_remain = k_delayed_work_remaining_get(&work_item_delayed);
+
+	k_sem_take(&sync_sema, K_FOREVER);
+	stop_time = k_cycle_get_32();
+	cycles_spent = stop_time - start_time;
+	ms_spent = (u32_t)k_cyc_to_ms_floor32(cycles_spent);
+
+	zassert_within(ms_spent, ms_remain, ms_delta, NULL);
+}
+
+/**
+ * @brief Test application can define workqueues without any limit
+ * @details
+ * - We can define any number of workqueus using a var of type struct k_work_q.
+ * - Define and initialize maximum possible real number of the workqueues
+ *   available according to the stack size.
+ * - Test defines and initializes max number of the workqueues and starts them.
+ * @ingroup kernel_workqueue_tests
+ */
+void test_workqueue_max_number(void)
```

Comment:
I still think this kind of requirement-chasing is silly, but this is as good a way to test this as any.

FWIW: this is likely to be fragile.  Setting MAX_WORK_Q_NUMBER high enough to be nontrivial means using a lot of memory for all those stacks.  We have a bunch of small devices in the CI test suite that are going to fail to link if you stuff too much junk into a single test.  This may need to be a kconfig and be set per-platform, or autodetected based on SRAM size.

Reply:
Yes, I agree with you. Seems that is why I received Shippable Error, because some devices can't support too many threads for work queues.

Reply:
@andyross Could you give an advice how it is better to implement? In configs documentation I can't find any SRAM sizes per platform.

Reply:
For now I just increased to CONFIG_MAX_THREAD_BYTES=3 to let Shippable passed and create 10 work queues for all platforms. I don't think if someone need more work queues in real application. Yes, that implementation is silly, because some platforms can handle more than 10 work queues... Maybe someone will want 50 work queues for his platform and, that is not tested and he will encounter with fatal error.